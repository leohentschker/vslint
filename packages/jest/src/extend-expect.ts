import * as fs from "node:fs";
import path from "node:path";
import type { ReviewRequest, ReviewResponse } from "@vslint/shared";
import axios, { type AxiosError, type AxiosResponse } from "axios";
import {
	DEFAULT_DESIGN_SNAPSHOT_DIR,
	DEFAULT_REVIEW_ENDPOINT,
} from "./constants";
import { getContentHash } from "./helpers";
import { getSnapshotIdentifier } from "./jest";
import { getLogger } from "./logging";
import { elementIsHTMLElement, getViewportSize } from "./render";
import { DEFAULT_RULES } from "./rules";
import {
	type DesignReviewMatcher,
	DesignReviewMatcherSchema,
	type DesignReviewRun,
	DesignReviewRunSchema,
} from "./types";
import { markdownToReviewResponse, reviewResponseToMarkdown } from "./markdown";

global.setImmediate =
	global.setImmediate ||
	((fn: TimerHandler, ...args: unknown[]) => global.setTimeout(fn, 0, ...args));

/**
 * Create a new jest matcher that exposes the `toPassDesignReview` method.
 * Expose this to your jest environment by running
 * expect.extend(extendExpectDesignReviewer({
 *   customStyles: ['./styles/globals.css'],
 *   model: {
 *     modelName: 'gpt-4o',
 *     key: process.env.OPENAI_API_KEY
 *   }
 * }));
 * @param {DesignReviewMatcher} args - Configuration for the design reviewer
 */
export const extendExpectDesignReviewer = (unsafeArgs: DesignReviewMatcher) => {
	const { data: args, error: extendValidationError } =
		DesignReviewMatcherSchema.safeParse(unsafeArgs);
	if (extendValidationError) {
		throw new Error(extendValidationError.message);
	}
	const { customStyles, snapshotsDir, reviewEndpoint, model, rules } = args;
	const designSnapshotsDir = snapshotsDir || DEFAULT_DESIGN_SNAPSHOT_DIR;
	for (const cssPath of customStyles) {
		if (!fs.existsSync(cssPath)) {
			throw new Error(
				`Could not find CSS file at path ${cssPath}. This file is required to correctly render your snapshots with your custom files.`,
			);
		}
	}

	// if the snapshots directory does not exist, log and create it
	if (!fs.existsSync(designSnapshotsDir)) {
		fs.mkdirSync(designSnapshotsDir, { recursive: true });
		getLogger().warn(
			`Created snapshots directory at path ${designSnapshotsDir}`,
		);
	}
	const stylesheets = customStyles.map((cssPath) =>
		fs.readFileSync(cssPath, "utf8"),
	);

	return {
		/**
		 * Make sure that the received element in `expect(...)` is an HTMLElement
		 * Run at different viewport sizes by using the `atSize` parameter
		 * Debug what's happening by using the `log` parameter to set different log levels
		 */
		async toPassDesignReview(
			received: unknown,
			unsafeParams?: DesignReviewRun,
		): Promise<jest.CustomMatcherResult> {
			const { data: params, error: runValidationError } =
				DesignReviewRunSchema.safeParse(unsafeParams);
			if (runValidationError) {
				return {
					pass: false,
					message: () =>
						runValidationError.errors.map((e) => e.message).join("\n"),
				};
			}

			const snapshotIdentifier = getSnapshotIdentifier(
				expect.getState(),
				params,
			);
			const logger = getLogger(params?.log);

			const snapshotPath = path.join(
				designSnapshotsDir,
				`${snapshotIdentifier}.md`,
			);
			logger.debug(`Snapshot path: ${snapshotPath}`);

			const snapshotFileExists = fs.existsSync(snapshotPath);
			if (snapshotFileExists)
				logger.debug("Snapshot file exists, pulling existing data");
			else logger.debug("Snapshot file does not exist, creating new snapshot");

			if (!elementIsHTMLElement(received)) {
				const errorMessage =
					"Received invalid value for element. Make sure that you pass a HTMLElement object into your expect call! This should look like:\nconst { container } = render(...);\nexpect(container).toPassDesignReview();";
				logger.error(errorMessage);
				return { pass: false, message: () => errorMessage };
			}

			const existingSnapshotFileContent = fs.existsSync(snapshotPath) && fs.readFileSync(snapshotPath, "utf8");
			const existingSnapshot: Partial<ReviewResponse> = existingSnapshotFileContent ? markdownToReviewResponse(existingSnapshotFileContent) : {};

			if (existingSnapshot.contentHash === getContentHash(received.outerHTML)) {
				logger.debug(
					`Snapshot ${snapshotPath} already exists and has the same content hash. Skipping review. To force a review, pass the option { forceReview: true } to your expect call.`,
				);
				return {
					pass: !!existingSnapshot.pass,
					message: () =>
						existingSnapshot.pass
							? "Snapshot passed design review"
							: `Snapshot failed design review.\n${existingSnapshot.explanation}\n\nTo override this, update the value of pass to true in ${snapshotPath}`,
				};
			}

			if (!model?.modelName || !model?.key) {
				throw new Error(
					process.env.CI
						? "vslint is running in a CI environment but the design snapshot does not exist for this test. vslint should never generate snapshots in CI as this could cause unintended model usage. Make sure to run this test locally and commit the results so re-review doesn't happen in CI."
						: "Component changed but can't re-run review as model name and key are not correctly set in `extendExpectDesignReviewer`. If you are setting the model key with an environment variable, make sure that `process.env.OPENAI_API_KEY` or `process.env.GEMINI_API_KEY` is set by running export OPENAI_API_KEY=...",
				);
			}

			const viewport = getViewportSize(params);
			logger.debug(`Viewport: ${JSON.stringify(viewport)}`);

			let response: AxiosResponse<ReviewResponse, ReviewRequest>;

			try {
				logger.debug("Sending request to review endpoint");
				const requestData: ReviewRequest = {
					content: received.outerHTML,
					stylesheets,
					rules: rules || DEFAULT_RULES,
					model,
					options: {
						viewport,
					},
					testDetails: {
						name: expect.getState().currentTestName || "",
					}
				};
				response = await axios.post(reviewEndpoint || DEFAULT_REVIEW_ENDPOINT, requestData);
			} catch (err) {
				const axiosError = err as AxiosError;
				logger.error(
					`Error while sending request to review endpoint: ${axiosError.message}`,
				);
				return {
					pass: false,
					message: () =>
						`Error while sending request to review endpoint: ${axiosError.message}`,
				};
			}
			const { content, explanation, pass, ...violations } = response.data;
			logger.debug(`Review result: ${JSON.stringify(violations, null, 2)}. Explanation: ${explanation}`);

			const imageSnapshotPath = path.join(
				designSnapshotsDir,
				`${snapshotIdentifier}.png`,
			);
			fs.writeFileSync(imageSnapshotPath, Buffer.from(response.data.content, "base64"));

			const reviewMarkdown = reviewResponseToMarkdown(response.data, imageSnapshotPath);
			if (!existingSnapshotFileContent || getContentHash(reviewMarkdown) !== getContentHash(existingSnapshotFileContent)) {
				logger.debug(`Snapshot ${snapshotPath} has changed, updating`);
				fs.writeFileSync(snapshotPath, reviewMarkdown);
			} else {
				logger.debug(
					`Snapshot ${snapshotPath} has not changed, skipping update`,
				);
			}

			const message = () =>
				pass
					? "Automated review successful"
					: `Design review failed: ${explanation}\n\nTo override this, update the value of pass to true in ${snapshotPath}`;
			return {
				message,
				pass,
			};
		},
	};
};
