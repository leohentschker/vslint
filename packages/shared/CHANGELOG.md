# @vslint/shared

## 0.0.8

### Patch Changes

- ac4bddf: Updates

## 0.0.7

### Patch Changes

- 77c8431: Fix build

## 0.0.6

### Patch Changes

- d9a9a19: Swapping fixture file format to markdown from JSON

## 0.0.5

### Patch Changes

- 85bb566: Updates available models to include gemini-1.5

## 0.0.4

### Patch Changes

- 38628f0: QOL improvements based on usage

## 0.0.3

### Patch Changes

- 31b0f4d: Fixing refinement on rules

## 0.0.2

### Patch Changes

- c61a70e: Exposes new package of shared schemas and types
