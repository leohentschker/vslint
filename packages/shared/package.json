{
  "name": "@vslint/shared",
  "version": "0.0.16",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "author": "Leo Hentschker <lhentschker@gmail.com>",
  "scripts": {
    "build": "tsup",
    "watch": "tsup --watch"
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": "./dist/index.mjs",
      "default": "./dist/index.js"
    }
  }
}
