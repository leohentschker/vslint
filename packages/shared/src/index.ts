export * from "./requests";
export * from "./rules";
export * from "./models";
export * from "./render";
export * from "./optional";
export * from "./tests";
export * from "./crypto";
export * from "./logging";
