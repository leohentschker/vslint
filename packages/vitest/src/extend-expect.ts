import * as fs from "node:fs";
import path from "node:path";
import {
  DEFAULT_RULES,
  type ReviewRequest,
  type ReviewResponse,
  getViewportSize,
} from "@vslint/shared";
import { DEFAULT_REVIEW_ENDPOINT } from "@vslint/shared";
import { getContentHash } from "@vslint/shared";
import { getSnapshotIdentifier } from "@vslint/shared";
import { getLogger } from "@vslint/shared";
import { elementIsHTMLElement } from "@vslint/shared";
import {
  type DesignReviewMatcher,
  DesignReviewMatcherSchema,
  type DesignReviewRun,
  DesignReviewRunSchema,
  JestSnapshotDataSchema,
} from "@vslint/shared";
import axios, { type AxiosError, type AxiosResponse } from "axios";
import Json5 from "json5";

const getExistingSnapshot = (matcherContext: jest.MatcherContext) => {
  const snapshotState = matcherContext.snapshotState; // Jest's snapshot state
  const currentTestName = matcherContext.currentTestName; // Current test name
  const count = snapshotState._counters.get(currentTestName) || 0;
  const snapshotKey = `${currentTestName} ${count + 1}`;
  let snapshotData = snapshotState._snapshotData[snapshotKey];
  if (!snapshotData) {
    return null;
  }
  snapshotData = snapshotData.replace(/^\\n/, "").trim();

  let parsedJson: unknown;
  try {
    parsedJson = Json5.parse(snapshotData);
  } catch (err) {
    return null;
  }

  const { data: parsedSnapshotData, error: parseError } =
    JestSnapshotDataSchema.safeParse(parsedJson);
  if (parseError) {
    return null;
  }
  return parsedSnapshotData;
};

const markSnapshotAsReviewed = (matcherContext: jest.MatcherContext) => {
  const snapshotState = matcherContext.snapshotState;
  const currentTestName = matcherContext.currentTestName;
  const count = snapshotState._counters.get(currentTestName) || 0;
  const snapshotKey = `${currentTestName} ${count + 1}`;
  snapshotState._counters.set(currentTestName, count + 1);
  snapshotState._uncheckedKeys.delete(snapshotKey);
};

global.setImmediate =
  global.setImmediate ||
  ((fn: TimerHandler, ...args: unknown[]) => global.setTimeout(fn, 0, ...args));

/**
 * Create a new jest matcher that exposes the `toPassDesignReview` method.
 * Expose this to your jest environment by running
 * expect.extend(extendExpectDesignReviewer({
 *   customStyles: ['./styles/globals.css'],
 *   model: {
 *     modelName: 'gpt-4o',
 *     key: process.env.OPENAI_API_KEY
 *   }
 * }));
 * @param {DesignReviewMatcher} args - Configuration for the design reviewer
 */
export const extendExpectDesignReviewer = (unsafeArgs: DesignReviewMatcher) => {
  const { data: args, error: extendValidationError } =
    DesignReviewMatcherSchema.safeParse(unsafeArgs);
  if (extendValidationError) {
    throw new Error(extendValidationError.message);
  }
  const {
    customStyles,
    reviewEndpoint,
    model,
    rules,
    strict: globalStrict,
  } = args;

  for (const cssPath of customStyles) {
    if (!fs.existsSync(cssPath)) {
      throw new Error(
        `Could not find CSS file at path ${cssPath}. This file is required to correctly render your snapshots with your custom files.`,
      );
    }
  }

  const stylesheets = customStyles.map((cssPath) =>
    fs.readFileSync(cssPath, "utf8"),
  );

  return {
    /**
     * Make sure that the received element in `expect(...)` is an HTMLElement
     * Run at different viewport sizes by using the `atSize` parameter
     * Debug what's happening by using the `log` parameter to set different log levels
     */
    async toPassDesignReview(
      received: unknown,
      unsafeParams?: DesignReviewRun,
    ): Promise<jest.CustomMatcherResult> {
      const { data: params, error: runValidationError } =
        DesignReviewRunSchema.safeParse(unsafeParams || {});

      if (runValidationError) {
        return {
          pass: false,
          message: () =>
            runValidationError.errors.map((e) => e.message).join("\n"),
        };
      }

      const logger = getLogger(params?.log);

      const strict = params?.strict ?? globalStrict;
      console.log("strict", strict);
      if (!elementIsHTMLElement(received)) {
        const errorMessage =
          "Received invalid value for element. Make sure that you pass a HTMLElement object into your expect call! This should look like:\nconst { container } = render(...);\nexpect(container).toPassDesignReview();";
        logger.error(errorMessage);
        return { pass: false, message: () => errorMessage };
      }

      const matcherContext = this as unknown as jest.MatcherContext;
      const existingSnapshot = getExistingSnapshot(matcherContext);

      const imageSnapshotFolder = path.join(
        path.dirname(matcherContext.snapshotState.snapshotPath),
        "vslint",
      );
      console.log("imageSnapshotFolder", imageSnapshotFolder);
      if (!fs.existsSync(imageSnapshotFolder)) {
        fs.mkdirSync(imageSnapshotFolder, { recursive: true });
      }
      const imageSnapshotPath = path.join(
        imageSnapshotFolder,
        `${getSnapshotIdentifier(matcherContext, params)}.png`,
      );

      // skip review if the snapshot already exists and the content hash matches
      if (existingSnapshot) {
        if (
          existingSnapshot.contentHash === getContentHash(received.outerHTML)
        ) {
          markSnapshotAsReviewed(matcherContext);
          let message: string;
          if (existingSnapshot.pass) {
            message =
              "Snapshot already exists, content hash matches, and the previous review passed";
          } else if (!strict) {
            message =
              "Snapshot already exists, content hash matches, previous review failed, but strict mode is disabled";
          } else {
            message = `Review failed and strict mode is enabled. Test failed rules: ${existingSnapshot.failedRules.join(", ")}. Set strict: false to skip review for this test. You can delete the snapshot file at ${matcherContext.snapshotState._snapshotPath} to force a new review. Review the snapshot at ${imageSnapshotPath} to see what failed.`;
          }
          return {
            pass: existingSnapshot.pass || !strict,
            message: () => message,
          };
        }
      }

      if (!args.model.key) {
        throw new Error(
          "No model key provided. Make sure to set the OPENAI_API_KEY environment variable via export OPENAI_API_KEY=<your-key>.",
        );
      }

      const viewport = getViewportSize(params);
      logger.debug(`Viewport: ${JSON.stringify(viewport)}`);

      let response: AxiosResponse<ReviewResponse, ReviewRequest>;

      try {
        logger.debug("Sending request to review endpoint");
        const requestData: ReviewRequest = {
          content: received.outerHTML,
          stylesheets,
          rules: rules || DEFAULT_RULES,
          model,
          options: {
            viewport,
          },
          testDetails: {
            name: expect.getState().currentTestName || "",
          },
        };
        response = await axios.post(
          reviewEndpoint || DEFAULT_REVIEW_ENDPOINT,
          requestData,
        );
      } catch (err) {
        const axiosError = err as AxiosError;
        logger.error(
          `Error while sending request to review endpoint: ${axiosError.message}`,
        );
        console.log("axiosError", axiosError);
        return {
          pass: false,
          message: () =>
            `Error while sending request to review endpoint: ${axiosError.message}`,
        };
      }
      const { explanation, pass, violations } = response.data;
      logger.debug(
        `Review result: ${JSON.stringify(violations, null, 2)}. Explanation: ${explanation}`,
      );

      const snapshotData = {
        contentHash: getContentHash(received.outerHTML),
        pass,
        failedRules: Object.entries(violations)
          .filter(([_, rule]) => rule.fail)
          .map(([ruleName]) => ruleName),
      };

      // Use Jest's built-in snapshot functionality and directly return
      expect(snapshotData).toMatchSnapshot();
      if (!pass) {
        getLogger().error(explanation);
      }

      const imageBuffer = Buffer.from(response.data.content, "base64");
      fs.writeFileSync(imageSnapshotPath, imageBuffer);
      console.log(imageSnapshotPath);
      console.log(matcherContext.snapshotState._snapshotPath);

      return { pass: pass || !strict, message: () => explanation || "" };
    },
  };
};
