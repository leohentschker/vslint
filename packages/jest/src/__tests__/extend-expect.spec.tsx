import { render, screen } from "@testing-library/react";
import { DEFAULT_REVIEW_TIMEOUT } from "@vslint/shared";
import React from "react";
import { extendExpectDesignReviewer } from "../extend-expect";

expect.extend(
  extendExpectDesignReviewer({
    customStyles: [],
    model: {
      modelName: "gpt-4o",
      key: process.env.OPENAI_API_KEY,
    },
  }),
);

test("React rendering works as expected", () => {
  render(<div>Hello</div>);
  const element = screen.getByText("Hello");
  expect(element).toBeInTheDocument();
});

test(
  "Snapshot tests work as expected",
  async () => {
    const { container } = render(
      <div>
        A server error has occurred. Please review your logs for more
        information.
      </div>,
    );
    await expect(container).toPassDesignReview({ atSize: "xs", strict: false });
  },
  DEFAULT_REVIEW_TIMEOUT,
);
