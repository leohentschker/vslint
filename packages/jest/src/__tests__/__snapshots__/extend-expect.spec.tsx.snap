// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests work as expected 1`] = `
{
  "contentHash": "1b7f124d21b942822d8fbad24d9b6206",
  "failedRules": [
    "missing-placeholders",
    "minimal-forms",
    "no-unconstrained-width",
  ],
  "pass": false,
}
`;
